addpkg(main/jellyfin): 10.10.7

NORALDM opened the pull request:
Closes https://github.com/termux/termux-packages/issues/9751
Supersedes https://github.com/termux/termux-packages/pull/23074
Had to go a bit overboard with the post_get_source step because multiple git urls in TERMUX_PKG_SRCURL don't seem to be supported. Most patches are taken from the termux packages repo except for skia-DEPS.patch, skia-gn-build.patch, and jellyfin-defaults.patch.

TomJo2000 commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,267 @@
+_FFMPEG_VERSION="7.0.2-9"
+TERMUX_PKG_SRCURL=(
+	"https://github.com/jellyfin/jellyfin/archive/refs/tags/v${TERMUX_PKG_VERSION}.tar.gz"
+	"https://github.com/jellyfin/jellyfin-ffmpeg/archive/refs/tags/v${_FFMPEG_VERSION}.tar.gz"
```
We can use a bit of substitution here to fix the URL for `jellyfin-ffmpeg`
```suggestion
TERMUX_PKG_SRCURL=(
	"https://github.com/jellyfin/jellyfin/archive/refs/tags/v${TERMUX_PKG_VERSION[0]}.tar.gz"
	"https://github.com/jellyfin/jellyfin-ffmpeg/archive/refs/tags/v${TERMUX_PKG_VERSION[1]%.*}-${TERMUX_PKG_VERSION[1]##*.}.tar.gz"
)
```

TomJo2000 commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,267 @@
+	local _idx;
+	local _sha256sums=();
+	local _error=0;
+	( [ "${#_GIT_URL[@]}" -eq "${#_GIT_COMMIT[@]}" ] && [ "${#_GIT_URL[@]}" -eq "${#_GIT_BRANCH[@]}" ] ) || termux_error_exit '_GIT_URL, _GIT_COMMIT and _GIT_BRANCH must be arrays of the same length'
```
Our build scripts are written for Bash, and arithmetic evaluation lets this be expressed a bit more nicely.
```suggestion
	(( ${#_GIT_URL[@]} == ${#_GIT_COMMIT[@]} && ${#_GIT_URL[@]} == ${#_GIT_BRANCH[@]} )) || termux_error_exit '_GIT_URL, _GIT_COMMIT and _GIT_BRANCH must be arrays of the same length'
```

TomJo2000 commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,267 @@
+	"git+https://github.com/ericsink/cb.git"
+	"git+https://github.com/ericsink/SQLitePCL.raw.git"
+	"git+https://github.com/mono/skia.git"
+	"git+https://github.com/jellyfin/jellyfin-web.git"
```
- 1.) There shouldn't be `.git` suffixes on those URLs.
- 2.) Consider using associcative arrays (hashmaps) if you need a secondary index, e.g. the project name.
`kitty`'s build script does something similar to this.
https://github.com/termux/termux-packages/blob/c1dd8cccf917885f12ecb3b81e202787ee43c16a/x11-packages/kitty/build.sh#L34-L51

TomJo2000 commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,267 @@
+		git fetch --depth 1 origin "${_GIT_COMMIT[_idx]}"
+		git -c advice.detachedHead=false checkout "${_GIT_COMMIT[_idx]}"
+		} >/dev/null
+		# we need a deterministic checksum, logs get in the way
```
This seems convoluted.
What's the goal here?

TomJo2000 commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,267 @@
+	elif [ "$TERMUX_ARCH" = "x86_64" ]; then
+		_target_cpu="x64"
+	elif [ "$TERMUX_ARCH" = "i686" ]; then
+		_target_cpu="x86"
```
Consider making this a `case` statement instead.

TomJo2000 reviewed (commented):
The `skiasharp` part of this build should probably be split out into a separate package that `jellyfin-server` can then depend on.

Same for `jellyfin-ffmpeg` probably.

Thank you for taking over this PR, it's quite a lot of work.

NORALDM commented:
Not sure why this closed on force pushing...

Asendaushoe commented:
Any process?

NORALDM commented:
Splitting the libraries into separate packages was part of my plan but it would require the rpath issue: #24271 to be resolved and I don't have much of an idea of how one could do that other than maintaining a list of CMakeLists that contain targets that need rpath to be set and using sed -i to insert add_linker_options directives after the required target()s, which is somewhat of a messy solution, which is why I was waiting for suggestions on how #24271 could be handled cleanly.

truboxl commented:
I felt the urge to diss a long explanation there but ended up not to since I doubt it will be helpful. You can wait infinitely or make do with adjustment here and move on.

truboxl commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,258 @@
+)
+TERMUX_PKG_SUGGESTS="jellyfin-ffmpeg"
+TERMUX_PKG_BUILD_DEPENDS="aspnetcore-targeting-pack-8.0, dotnet-targeting-pack-8.0, libcairo, pango, libjpeg-turbo, giflib, librsvg"
+TERMUX_PKG_DEPENDS="libc++, fontconfig, aspnetcore-runtime-8.0, dotnet-host, dotnet-runtime-8.0, sqlite, libexpat, libpng, libwebp, freetype, ffmpeg"
```
```suggestion
TERMUX_PKG_DEPENDS="libc++, fontconfig, aspnetcore-runtime-8.0, dotnet-host, dotnet-runtime-8.0, sqlite, libexpat, libpng, libwebp, freetype"
```

truboxl commented on packages/jellyfin-server/jellyfin-ffmpeg.subpackage.sh:
```diff
@@ -0,0 +1,4 @@
+TERMUX_SUBPKG_DESCRIPTION="FFmpeg for Jellyfin with custom extensions and enhancements"
+TERMUX_SUBPKG_DEPENDS="chromaprint, libdrm, zstd"
```
```suggestion
TERMUX_SUBPKG_DEPENDS="chromaprint, fontconfig, freetype, fribidi, game-music-emu, harfbuzz, libaom, libandroid-glob, libandroid-stub, libass, libbluray, libbz2, libdav1d, libdrm, libgnutls, libiconv, liblzma, libmp3lame, libopencore-amr, libopenmpt, libopus, librav1e, libsoxr, libsrt, libssh, libtheora, libv4l, libvidstab, libvmaf, libvo-amrwbenc, libvorbis, libvpx, libwebp, libx264, libx265, libxml2, libzimg, libzmq, littlecms, ocl-icd, rubberband, svt-av1, xvidcore, zlib, zstd"
```

NORALDM replied:
I'll add the deps to the subpackage but chromaprint pulls in ffmpeg anyways for FFT

truboxl commented on packages/jellyfin-server/build.sh:
```diff
@@ -0,0 +1,258 @@
+	popd
+}
+termux_step_pre_configure() {
+	termux_setup_dotnet; termux_setup_gn; termux_setup_nodejs
```
```suggestion
	TERMUX_PKG_DEPENDS+=", jellyfin-ffmpeg"
	termux_setup_dotnet; termux_setup_gn; termux_setup_nodejs
```

truboxl reviewed (commented):
Since upstream feels so strongly about using its own jellyfin-ffmpeg instead of system ffmpeg then just use jellyfin-ffmpeg. Preferably prioritise disk space usage than pulling unneccessary packages.

You can workaround the issue like so: (I havent check the dependency list yet)

NORALDM commented:
Probably out of the scope of this PR, but (aping how Alpine Linux does it) chromaprint could be split into chromaprint and libchromaprint by linking with fftw instead of ffmpeg for the FFT, which would make ffmpeg a dependency of the main chromaprint package only (for fpcalc), and jellyfin-ffmpeg could dep on libchromaprint to avoid pulling in ffmpeg.

NORALDM commented:
Woops, didn't get the failed re-rerun in my notifs. I'm going to rebase if that's ok with you @truboxl

truboxl commented:
I will take a look splitting chromaprint

It is still unacceptable large to pull 2 copies of ffmpeg

truboxl commented:
Ok, you can open a new PR

NORALDM commented:
> Ok, you can open a new PR

Chromaprint PR has been merged so this Jellyfin should just pull in its own ffmpeg

robertkirkman commented:
I tested this, and this error occurs for me,

```
~/output $ export LD_LIBRARY_PATH="$PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native"
~/output $ file $PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/libhostfxr.so
/data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/libhostfxr.so: symbolic link to ../../../../../../host/fxr/8.0.21/libhostfxr.so
~/output $ file $PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so
/data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, for Android 24, built by NDK r28c (13676358), BuildID[sha1]=69c90b3fdd8bc3c2489b5888e948e06ecc120c06, stripped
~/output $ realpath /data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so
/data/data/com.termux/files/usr/lib/dotnet/host/fxr/8.0.21/libhostfxr.so
~/output $ jellyfin 
You must install .NET to run this application.

App: /data/data/com.termux/files/usr/lib/jellyfin/jellyfin
Architecture: arm64
App host version: 8.0.21
.NET location: Not found

Learn more:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-bionic-arm64&os=linux-bionic&apphost_version=8.0.21
Failed to resolve libhostfxr.so [not found]. Error code: 0x80008083
~/output $ export LD_LIBRARY_PATH="$PREFIX/lib/dotnet/host/fxr/8.0.21"
~/output $ jellyfin 
You must install .NET to run this application.

App: /data/data/com.termux/files/usr/lib/jellyfin/jellyfin
Architecture: arm64
App host version: 8.0.21
.NET location: Not found

Learn more:
https://aka.ms/dotnet/app-launch-failed

Download the .NET runtime:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-bionic-arm64&os=linux-bionic&apphost_version=8.0.21
Failed to resolve libhostfxr.so [not found]. Error code: 0x80008083
~/output $ 
```

- I assume this is related to the problem discussed above regarding https://github.com/termux/termux-packages/issues/24271

however unfortunately I am not experienced with C# and I don't know how to fix it.

quanuanc commented:
> I tested this, and this error occurs for me,
> 
> ```
> ~/output $ export LD_LIBRARY_PATH="$PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native"
> ~/output $ file $PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/libhostfxr.so
> /data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/libhostfxr.so: symbolic link to ../../../../../../host/fxr/8.0.21/libhostfxr.so
> ~/output $ file $PREFIX/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so
> /data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so: ELF 64-bit LSB shared object, ARM aarch64, version 1 (SYSV), dynamically linked, for Android 24, built by NDK r28c (13676358), BuildID[sha1]=69c90b3fdd8bc3c2489b5888e948e06ecc120c06, stripped
> ~/output $ realpath /data/data/com.termux/files/usr/lib/dotnet/packs/Microsoft.NETCore.App.Runtime.linux-bionic-arm64/8.0.21/runtimes/linux-bionic-arm64/native/../../../../../../host/fxr/8.0.21/libhostfxr.so
> /data/data/com.termux/files/usr/lib/dotnet/host/fxr/8.0.21/libhostfxr.so
> ~/output $ jellyfin 
> You must install .NET to run this application.
> 
> App: /data/data/com.termux/files/usr/lib/jellyfin/jellyfin
> Architecture: arm64
> App host version: 8.0.21
> .NET location: Not found
> 
> Learn more:
> https://aka.ms/dotnet/app-launch-failed
> 
> Download the .NET runtime:
> https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-bionic-arm64&os=linux-bionic&apphost_version=8.0.21
> Failed to resolve libhostfxr.so [not found]. Error code: 0x80008083
> ~/output $ export LD_LIBRARY_PATH="$PREFIX/lib/dotnet/host/fxr/8.0.21"
> ~/output $ jellyfin 
> You must install .NET to run this application.
> 
> App: /data/data/com.termux/files/usr/lib/jellyfin/jellyfin
> Architecture: arm64
> App host version: 8.0.21
> .NET location: Not found
> 
> Learn more:
> https://aka.ms/dotnet/app-launch-failed
> 
> Download the .NET runtime:
> https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=arm64&rid=linux-bionic-arm64&os=linux-bionic&apphost_version=8.0.21
> Failed to resolve libhostfxr.so [not found]. Error code: 0x80008083
> ~/output $ 
> ```
> 
>  * I assume this is related to the problem discussed above regarding [[Bug]: Some executables and libraries in the dotnet packages need RUNPATH to be set #24271](https://github.com/termux/termux-packages/issues/24271)
> 
> however unfortunately I am not experienced with C# and I don't know how to fix it.

Try this:
```shell
# first install dotnet
pkg install dotnet-host
pkg install dotnet-runtime-8.0

# then set DOTNET_ROOT and run jellyfin
export DOTNET_ROOT=$PREFIX/lib/dotnet
jellyfin
```

robertkirkman commented:
Thank you, I didn't know about that, that actually does work and it's working now!

alexytomi commented:
Another thing to add, on some phones, dotnet just crashes with an OOM, run `export DOTNET_GCHeapHardLimit=0x1F400000` (this is about 528mb) if needed

NORALDM commented:
> ```
> [06:35:29] [FTL] [1] Main: Unhandled Exception
> System.ObjectDisposedException: Cannot access a disposed object.
> Object name: 'LoggerFactory'.
>    at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
>    at Jellyfin.Server.Migrations.JellyfinMigrationService.MigrateStepAsync(JellyfinMigrationStageTypes stage, IServiceProvider serviceProvider)
>    at Jellyfin.Server.Migrations.JellyfinMigrationService.MigrateStepAsync(JellyfinMigrationStageTypes stage, IServiceProvider serviceProvider)
>    at Jellyfin.Server.Program.ApplyStartupMigrationAsync(ServerApplicationPaths appPaths, IConfiguration startupConfig)
>    at Jellyfin.Server.Program.StartApp(StartupOptions options)
>    at Jellyfin.Server.Program.<Main>(String[] args)
> ~ $ 
> ```
> 
> do you happen to know what might be wrong with this?

Seems to be an upstream bug, https://github.com/jellyfin/jellyfin/issues/15571

truboxl commented:
> Force pushing to fix weird broken symlinks in `$PREFIX/lib/dotnet` in the last CI artifacts

To counter this I suggest adding `lib/dotnet` to `TERMUX_PKG_RM_AFTER_INSTALL`

NORALDM commented:
> Unfortunately, with this build I am still seeing the same error:
> 
> ```
> [19:25:28] [INF] [9] Jellyfin.Server.Migrations.JellyfinMigrationService: There are 2 migrations for stage PreInitialisation.
> [19:25:28] [INF] [9] InternalCodeMigration: Perform migration 20250420030000_MigrateEncodingOptions
> [19:25:28] [INF] [9] InternalCodeMigration: Migration 20250420030000_MigrateEncodingOptions was successfully applied
> [19:25:28] [FTL] [1] Main: Unhandled Exception
> System.ObjectDisposedException: Cannot access a disposed object.
> Object name: 'LoggerFactory'.
>    at Microsoft.Extensions.Logging.LoggerFactory.CreateLogger(String categoryName)
>    at Jellyfin.Server.Migrations.JellyfinMigrationService.MigrateStepAsync(JellyfinMigrationStageTypes stage, IServiceProvider serviceProvider)
>    at Jellyfin.Server.Migrations.JellyfinMigrationService.MigrateStepAsync(JellyfinMigrationStageTypes stage, IServiceProvider serviceProvider)
>    at Jellyfin.Server.Program.ApplyStartupMigrationAsync(ServerApplicationPaths appPaths, IConfiguration startupConfig)
>    at Jellyfin.Server.Program.StartApp(StartupOptions options)
>    at Jellyfin.Server.Program.<Main>(String[] args)
> ~/bin $ 
> ```

I think this happens due to an error having happened during migration, and similar errors are reported in the Jellyfin forum with a suggested fix being moving/renaming migrations.xml, found in `~/.config/jellyfin` or `~/.local/share/jellyfin/config` (the former on my device). If that fails your library may be corrupted and you would need to move/rename `~/.local/share/jellyfin/data/library.db`.

robertkirkman reviewed (approved):
Oh, good thanks!

I tried that suggestion and it worked, now the `jellyfin` command is working!

Since the package has not yet been in Termux, and the older version that caused the problem wasn't in Termux, I don't think it will be necessary to worry that this might happen to someone else.

Is this ready?

truboxl commented:
Please apply the ffmpeg patch for Android 15 Mediacodec fix

truboxl commented:
If so, I dont see a need to apply patch for reorder the priority of codec selection

NORALDM commented:
The libandroid dependency was due to the android_camera device but I don't think Jellyfin would ever need that, so I added `--disable-indev=android_camera` and removed the libandroid-stub dependency from jellyfin-ffmpeg in this force push.

robertkirkman commented:
Ok, now it is ready, right?

NORALDM commented:
It's mostly ready, but there's a behavioural difference between this and normal Jellyfin introduced in the new update probably due to runtime differences in Termux .NET and normal .NET. Whenever a library scan encounters an empty folder, it posts a warning to the log without crashing, doesn't seem to cause any visible side effects, and the scan completes normally. This doesn't happen on jellyfin on Alpine Linux on proot-distro.

```
[13:15:51] [ERR] [23] MediaBrowser.Providers.Folders.FolderMetadataService: Error in Dynamic Image Provider
System.ArgumentException: The type ImageInfoImageType[] does not represent a sequence at System.SharedTypeExtensions.GetSequenceType(Type type) [...]
```
